### PR TITLE
177 452 date module enhancement

### PR DESCRIPTION
**Pull request by thomas-rcv**
- Comment function for participants  #452 
- Enable further guests for participants of dates  #177

### CODE REVIEW

File: adm_program/modules/dates/dates.php [hunk 1]
```diff
@@ -33,8 +33,8 @@
 unset($_SESSION['dates_request']);
 
 // Initialize and check the parameters
-$disableAdditionalGuests   = 'disabled';
-$disableComments           = 'disabled';
+$disableAdditionalGuests   = (int) 2;
```
Comment (ximex) [inline adm_program/modules/dates/dates.php]: You don't have to cast a static int to an int

File: adm_program/modules/dates/dates.php [hunk 2]
```diff
@@ -33,8 +33,8 @@
...
+$disableComments           = (int) 2;
```
Comment (ximex) [inline adm_program/modules/dates/dates.php]: You don't have to cast a static int to an int

File: adm_program/modules/dates/dates.php [hunk 3]
```diff
@@ -618,8 +615,8 @@ function submitParticipationForm(datId, mode){
                         </div>');
                 }
                 // Reset flags and parameters
-                $disableAdditionalGuests    = 'disabled';
-                $disableComments            = 'disabled';
+                $disableAdditionalGuests    = (int) 2;
```
Comment (ximex) [inline adm_program/modules/dates/dates.php]: You don't have to cast a static int to an int

File: adm_program/modules/dates/dates.php [hunk 4]
```diff
@@ -618,8 +615,8 @@ function submitParticipationForm(datId, mode){
...
+                $disableComments            = (int) 2;
```
Comment (ximex) [inline adm_program/modules/dates/dates.php]: You don't have to cast a static int to an int

File: adm_program/system/classes/participants.php [hunk 5]
```diff
@@ -99,25 +99,33 @@ public function getCount($rolId = 0)
         while ($row = $membersStatement->fetch())
         {
             $numParticipants[] = array(
-                'member' => (int) $row['mem_usr_id'],
-                'leader' => (bool) $row['mem_leader']
+                'member'            => (int) $row['mem_usr_id'],
+                'leader'            => (bool) $row['mem_leader'],
+                'count_guests' => (int) $row['mem_count_guests']
```
Comment (ximex) [inline adm_program/system/classes/participants.php]: Please align the `=>`

File: adm_program/modules/dates/dates.php [hunk 6]
```diff
@@ -318,6 +321,21 @@
         $dateElements        = array();
         $participantsArray   = array();
 
+        // If extended options for participation are allowed then use a modal form instead the dropdown button
+        If ((int) $date->getValue('dat_allow_comments') === 1 || (int) $date->getValue('dat_additional_guests') === 1 )
```
Comment (ximex) [inline adm_program/modules/dates/dates.php]: `If` in lowercase

File: adm_program/modules/dates/dates.php [hunk 7]
```diff
@@ -133,6 +132,13 @@
             window.open("'.ADMIDIO_URL.FOLDER_MODULES.'/dates/dates.php?view_mode=print&view=' . $getView . '&mode=' . $getMode . '&headline=' . $getHeadline . '&cat_id=' . $getCatId . '&id=' . $getId . '&date_from=' . $dates->getParameter('dateStartFormatEnglish') . '&date_to=' . $dates->getParameter('dateEndFormatEnglish') . '", "_blank");
         });', true);
 
+    $page->addJavascript('
+        function submitParticipationForm(datId, mode){
```
Comment (ximex) [inline adm_program/modules/dates/dates.php]: Please change to: `function submitParticipationForm(datId, mode) {` (whitespace in front of `{`)

File: adm_program/modules/dates/dates.php [hunk 8]
```diff
@@ -581,21 +586,25 @@
...
                                             <img src="'.THEME_URL.'/icons/ok.png" alt="' . $gL10n->get('DAT_ATTEND') . '" title="' . $gL10n->get('DAT_ATTEND') . '"/>' . $gL10n->get('DAT_ATTEND') . '
                                         </a>
-                                        <a class="btn btn-default" href="'.ADMIDIO_URL.FOLDER_MODULES.'/dates/dates_function.php?mode=7&amp;dat_id=' . $dateId . '"' . $disableStatusTentative . '>
+                                        <a class="btn btn-default" onclick="submitParticipationForm(' . $dateId .', 7)"' . $disableStatusTentative . '>
                                             <img src="'.THEME_URL.'/icons/help_violett.png" alt="' . $gL10n->get('DAT_USER_TENTATIVE') . '" title="' . $gL10n->get('DAT_USER_TENTATIVE') . '"/>' . $gL10n->get('DAT_USER_TENTATIVE') . '
                                         </a>
-                                        <a class="btn btn-default" href="'.ADMIDIO_URL.FOLDER_MODULES.'/dates/dates_function.php?mode=4&amp;dat_id=' . $dateId . '">
+                                        <a class="btn btn-default" onclick="submitParticipationForm(' . $dateId .', 4)"">
```
Comment (ximex) [inline adm_program/modules/dates/dates.php]: There is a `"` to much

File: adm_program/modules/dates/dates.php [hunk 9]
```diff
@@ -581,21 +586,25 @@
...
+                                            <label for="additonal_guests">' .$gL10n->get('LST_SEAT_AMOUNT'). ':
+                                                <input name="additonal_guests" alt="additonal_guests" class="form-control" value="' .$row['additional_guests'] . '"' . $disableAdditionalGuests . '>
+                                            </label>
+                                            <input type="hidden" name="dat_id" value="' . $dateId .'">
+                                            <input type="hidden" name="mode" value="">
+                                        </form>
                                     </div>
                                     <div class="modal-footer">
-                                        <a class="btn btn-default" href="'.ADMIDIO_URL.FOLDER_MODULES.'/dates/dates_function.php?mode=3&amp;dat_id=' . $dateId . '"' . $disableStatusAttend . '>
+                                        <a class="btn btn-default" onclick="submitParticipationForm(' . $dateId .', 3)"' . $disableStatusAttend . ' >
```
Comment (ximex) [inline adm_program/modules/dates/dates.php]: unnecessary space in front of `>`

File: adm_program/installation/db_scripts/update_3_3.xml [hunk 10]
```diff
@@ -10,6 +10,8 @@
     <step id="80">ALTER TABLE %PREFIX%_roles DROP COLUMN rol_visible</step>
     <step id="90">ComponentUpdate::updateStepEventCategory</step>
     <step id="100">ComponentUpdate::updateStepSetParticipantsApprovalStates</step>
+    <step id="110">ALTER TABLE %PREFIX%_dates ADD COLUMN dat_allow_comments integer not null default '0'</step>
```
Comment (thomas-rcv) [inline adm_program/installation/db_scripts/update_3_3.xml]: changed

File: adm_program/installation/db_scripts/update_3_3.xml [hunk 11]
```diff
@@ -10,6 +10,8 @@
...
+    <step id="120">ALTER TABLE %PREFIX%_dates ADD COLUMN dat_allow_guests integer not null default '0'</step>
```
Comment (thomas-rcv) [inline adm_program/installation/db_scripts/update_3_3.xml]: changed

File: adm_program/modules/dates/dates.php (hunk 1 above)
Comment (thomas-rcv) [inline adm_program/modules/dates/dates.php]: Forgot to remove. Sorry. 
Needed this for testing with basic html form implementation.
After changing to HtmlForm with property => FIELD_HIDDEN is not nescessary anymore.

File: adm_program/modules/dates/dates.php (hunk 2 above)
Comment (thomas-rcv) [inline adm_program/modules/dates/dates.php]: removed

File: adm_program/modules/dates/dates.php (hunk 7 above)
Comment (thomas-rcv) [inline adm_program/modules/dates/dates.php]: typo fixed

File: adm_program/modules/dates/dates.php (hunk 6 above)
Comment (thomas-rcv) [inline adm_program/modules/dates/dates.php]: typo fixed

File: adm_program/modules/dates/dates.php (hunk 3 above)
Comment (thomas-rcv) [inline adm_program/modules/dates/dates.php]: removed

File: adm_program/modules/dates/dates.php (hunk 4 above)
Comment (thomas-rcv) [inline adm_program/modules/dates/dates.php]: removed

File: adm_program/modules/dates/dates.php (hunk 9 above)
Comment (thomas-rcv) [inline adm_program/modules/dates/dates.php]: typo fixed

File: adm_program/modules/dates/dates.php (hunk 8 above)
Comment (thomas-rcv) [inline adm_program/modules/dates/dates.php]: typo fixed

File: adm_program/system/classes/participants.php (hunk 5 above)
Comment (thomas-rcv) [inline adm_program/system/classes/participants.php]: typo fixed

**Comment by thomas-rcv** [comment]
@ fasse: 
Issues are fixed. Most depending stuff of html output was simply the change from type navbar to default. But this change occurs other structual problems. Please take not of the upcoming email to you and Ximex.
We should discuss this. The textarea is now hidden as well. This was not simply possible. Please have look of the dedending bug commit.

**Comment by thomas-rcv** [comment]
Should be finished now.

File: adm_program/modules/dates/dates_function.php [hunk 12]
```diff
@@ -575,9 +575,37 @@
     $member = new TableMembers($gDb);
     $member->readDataByColumns(array('mem_rol_id' => $date->getValue('dat_rol_id'), 'mem_usr_id' => $gCurrentUser->getValue('usr_id')));
     $member->setValue('mem_comment', $getUserComment);
-    $member->setValue('mem_count_guests', $getAdditionalGuests);
+    // If participation limit is set
+    if ($date->getValue('dat_max_members') > 0)
+    {
+        // First store possible previous guests assignment and reset value
+        $currentGuests = $member->getValue('mem_count_guests');
+        $member->setValue('mem_count_guests', NULL);
```
Comment (Fasse) [inline adm_program/modules/dates/dates_function.php]: Why do you save NULL to the database and therefore delete the current value in database?
What if the next steps failed? Than you have lost your stored value.

Comment (thomas-rcv) [inline adm_program/modules/dates/dates_function.php]: You are right. This is non sense. This was old code from first implementation before rework.
Deleted it.

**Comment by papierkorb27** [comment]
If I understood correctly the event leader shouldnt be able to set this status unfortunately: https://www.admidio.org/forum/viewtopic.php?f=11&t=7375

**Comment by thomas-rcv** [comment]
@fasse: 
This was not the plan.
Doing so, you will never have a correct overview in your event plans.
What if you set automatically the status to "approved will participate" and the user doesn´t react if he will not participate? In my opinion this would be not a good way.

We might think about an additional Feature for leaders to set the states for the users in future, but definitly not automatically while assignment.
Personally I also can´t see any benefit having this possibilty. If a users tells you that he will not participate just remove from the participation list. 

Of course, we can enhance in future versions. I think we have a lot of new features now, and for me  the next step is definitly the beta test and the feedback from our users with bugs and the look and feel.
Adjustments and changes are possible everytime.

**Comment by Fasse** [comment]
We should discuss this in the forum. You can merge this feature request.

**Pull request merged**